### ISSUE TITLE
ocp4upc failing in errata mode due to next release channel selection

### ISSUE THREAD

**Issue by palonsoro** [pamoedom/ocp4upc#4]
I am facing the following failure:

It seems that, by default, ocp4upc tries to get the release update paths on the next release channel. However, the `stable-4.5` and `fast-4.5` channels are ready but contain no edges, so any 4.4 errata path fails.

Sample execution:
```
$ ocp4upc.sh 4.4.3
[INFO] Checking prerequisites (curl jq dot)... [OK] 
[INFO] Errata provided (4.x.z mode), targeting '4.5' channels for upgrade path generation.
[INFO] Checking if '4.4.3' (amd64) is a valid release... [OK] 
[WARN] Skipping channel 'stable-4.5_amd64', it's void.
[WARN] Skipping channel 'fast-4.5_amd64', it's void.
[ERROR] There are no channels to process. Aborting execution.
```

Furthermore, even if we are not in the corner case of the latest release with the next release channels blank, this also prevents to show any upgrade paths for a 4.y.z version that is not on the channels of the 4.(y+1) version. 

Example:
```
$ ocp4upc.sh 4.3.0
[INFO] Checking prerequisites (curl jq dot)... [OK] 
[INFO] Errata provided (4.x.z mode), targeting '4.4' channels for upgrade path generation.
[INFO] Checking if '4.3.0' (amd64) is a valid release... [OK] 
[WARN] Discarding channel 'fast-4.4_amd64', it doesn't differ from 'stable-4.4_amd64'.
[WARN] Skipping channel 'stable-4.4_amd64', version '4.3.0' not found.
[ERROR] Version '4.3.0' not found (or not upgradable) within '4.4' channels. Aborting execution.
[INFO] TIP: try only with target release (e.g. 4.4) to check available versions.
```

Maybe simplest solution would be to add a flag to allow searching upgrade paths in errata mode staying in the current release. What do you think?

**Comment by pamoedom** [comment]
Hi Pablo, thanks for your feedback.

The behavior you are seeing in the corner case of 4.4 channel is expected, the script is designed to generate the next minor channel upgrade path in errata mode (4.x.z) and therefore if next minor channels are empty (like 4.5), the script will abort the execution.

For those cases, there is also the other basic mode (4.x) which will extract the whole channel graph of the same minor release, I know this is not ideal if you are looking for an specific upgrade path in a big channel, but changing this behavior in the script will require multiple workflow modifications.

Having said this, let me think if I can include something simple that can allow user interaction to decide on those corner cases instead of adding flags, I prefer to keep input arguments as simple as possible.

Regards.

**Comment by palonsoro** [comment]
Maybe always generating the paths for both the N and N+1 channel?

**Comment by pamoedom** [comment]
I don't like the script to generate non-requested extra output nor to deviate from the original goal of next-minor upgrade path tool, however, I've implemented the following within latest v2.5 to cover the corner case in the following manner:

~~~
$ ocp4upc 4.4.3
[INFO] Checking prerequisites (curl jq dot)... [OK] 
[INFO] Errata provided (4.x.z mode), targeting '4.5' channels for upgrade path generation.
[INFO] Checking if '4.4.3' (amd64) is a valid release... [OK] 
[WARN] Skipping channel 'stable-4.5_amd64', it's empty.
[WARN] Skipping channel 'fast-4.5_amd64', it's empty.
[WARN] It seems you are targeting void channels, do you want to generate same-minor upgrade path instead? (y/N):y
[INFO] Result exported as 'stable-4.4_amd64_20200708.svg'
[INFO] Result exported as 'fast-4.4_amd64_20200708.svg'
~~~

I've decided to do it only as a corner case when targeting void channels because for the other cases where this situation could apply, we already have the command `oc adm upgrade` that shows all the upgrade paths within the same minor version, for example:

~~~
$ oc adm upgrade
Cluster version is 4.2.7

Updates:
VERSION IMAGE
4.2.8   quay.io/openshift-release-dev/ocp-release@sha256:4bf307b98beba4d42da3316464013eac120c6e5a398646863ef92b0e2c621230
4.2.9   quay.io/openshift-release-dev/ocp-release@sha256:f28cbabd1227352fe704a00df796a4511880174042dece96233036a10ac61639
4.2.10  quay.io/openshift-release-dev/ocp-release@sha256:dc2e38fb00085d6b7f722475f8b7b758a0cb3a02ba42d9acf8a8298a6d510d9c
4.2.12  quay.io/openshift-release-dev/ocp-release@sha256:77ade34c373062c6a6c869e0e56ef93b2faaa373adadaac1430b29484a24d843
4.2.13  quay.io/openshift-release-dev/ocp-release@sha256:782b41750f3284f3c8ee2c1f8cb896896da074e362cf8a472846356d1617752d
4.2.14  quay.io/openshift-release-dev/ocp-release@sha256:3fabe939da31f9a31f509251b9f73d321e367aba2d09ff392c2f452f6433a95a
4.2.16  quay.io/openshift-release-dev/ocp-release@sha256:e5a6e348721c38a78d9299284fbb5c60fb340135a86b674b038500bf190ad514
4.2.18  quay.io/openshift-release-dev/ocp-release@sha256:283a1625e18e0b6d7f354b1b022a0aeaab5598f2144ec484faf89e1ecb5c7498
4.2.36  quay.io/openshift-release-dev/ocp-release@sha256:449b9b839d2cdf33ff2a494a5863d24dc1b436f995c286d8f8d58821ae293b82
~~~

Regards.

**Comment by pamoedom** [comment]
Hi again Pablo, I've uploaded version 2.6 with a new mode that allows also to generate same-minor upgrade paths, please take a look, I'm closing this issue ;)

~~~
-------------------------------------------------------------------
OCP4 Upgrade Paths Checker (stable fast) v2.6

Usage:
/home/pamoedo/bin/ocp4upc source_version [arch]

Source Version:
4.x        Extract default graphs using same-minor channels, e.g. '4.2'
4.x.z      Generate upgrade paths using next-minor channels, e.g. '4.2.26'
4.x.z.     Generate upgrade paths using same-minor channels, e.g. '4.2.26.'

Arch (optional):
amd64      x86_64 (default)
s390x      IBM System/390
ppc64le    POWER8 little endian
-------------------------------------------------------------------
~~~

Regards.
